services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,428 @@
+import org.eclipse.hono.service.tenant.TenantService;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.AuthenticationConstants;
+import org.eclipse.hono.util.RegistrationConstants;
+import org.eclipse.hono.util.RegistryManagementConstants;
+import org.eclipse.hono.util.TenantResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.opentracing.Span;
+import io.opentracing.noop.NoopSpan;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.mongo.FindOptions;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.UpdateOptions;
+
+/**
+ * An implementation of the tenant service and the tenant management service
+ * that uses a mongodb database to store tenants.
+ * <p>
+ * On startup this adapter tries to find the tenant collection, if not found, it gets created.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/tenant/">Tenant API</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/management/">Device Registry Management API</a>
+ */
+public final class MongoDbBasedTenantService implements TenantService, TenantManagementService, Lifecycle {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoDbBasedTenantService.class);
+
+    private static final int INDEX_CREATION_MAX_RETRIES = 3;
+    private final MongoClient mongoClient;
+    private final MongoDbCallExecutor mongoDbCallExecutor;
+    private final MongoDbBasedTenantsConfigProperties config;
+
+    /**
+     * Creates a new service for configuration properties.
+     *
+     * @param vertx The vert.x instance to run on.
+     * @param mongoClient The client for accessing the Mongo DB instance.
+     * @param config The properties for configuring this service.
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     */
+    public MongoDbBasedTenantService(
+            final Vertx vertx,
+            final MongoClient mongoClient,
+            final MongoDbBasedTenantsConfigProperties config) {
+
+        Objects.requireNonNull(vertx);
+        Objects.requireNonNull(mongoClient);
+        Objects.requireNonNull(config);
+
+        this.mongoClient = mongoClient;
+        this.mongoDbCallExecutor = new MongoDbCallExecutor(vertx, mongoClient);
+        this.config = Objects.requireNonNull(config);
+    }
+
+    @Override
+    public Future<Void> start() {
+        return CompositeFuture.all(
+                mongoDbCallExecutor.createCollectionIndex(config.getCollectionName(),
+                        new JsonObject().put(RegistrationConstants.FIELD_PAYLOAD_TENANT_ID, 1),
+                        new IndexOptions().unique(true), INDEX_CREATION_MAX_RETRIES),
+                mongoDbCallExecutor.createCollectionIndex(config.getCollectionName(),
```

Comment:
Sorry, but I don't know much about MongoDB. I think it would help documenting a little bit, what this is expected to do.

Reply:
I was thinking more about "why" you do things like that. Because I already assume that `new IndexOptions().unique(true)` will create a "unique index". So I do understand "what" you do, the question "why" and "why this way".